### ISSUE TITLE
stream.stop() breaks when async=True

### ISSUE THREAD

**Issue by jackiosport** [liampauling/betfair#165]
When calling `stream.stop()` when `async=True`, I get the following error:

`Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/betfairlightweight/streaming/betfairstream.py", line 191, in _read_loop
    received_data_raw = self._receive_all()
  File "/usr/local/lib/python2.7/dist-packages/betfairlightweight/streaming/betfairstream.py", line 223, in _receive_all
    raise SocketError('Connection closed by server')
SocketError: Connection closed by server`

This is because the threads calling `stream._read_loop` don't have time to see that `stream._running` has been set to `False`, thus are still trying to read from the socket, even though the socket has been closed by the `stream.stop()`.

A dirty solution to this is to do:
`stream._running = False`
`time.sleep(5) # this gives time for _read_loop threads to see they need to stop reading from the socket.` 
`stream.stop()`
However, I think I nicer solution would to stop the threads directly in the stream.stop() function before the socket is closed. 
I can add this fix and test locally. Would you want me to merge this into master?

Cheers,
Jack

**Comment by liampauling** [comment]
Hi Jack, how are you getting this error as they only way I can recreate is by subscribing, starting, stoping and then starting again which shouldn't be done (should always subscribe before starting).

**Comment by jackiosport** [comment]
For me I just need to start the stream (in async). Then stop and I get this error.

**Comment by liampauling** [comment]
As in you don't subscribe before running start? Can you share that part of the code?

**Comment by jackiosport** [comment]
`import os
import logging
import queue
import time

import betfairlightweight
from betfairlightweight.filters import (
    streaming_market_filter,
    streaming_market_data_filter,
)



logging.basicConfig(level=logging.INFO)

class footballStreamer(object):

    def __init__(self,un,pw,key,path_to_certs,q,freq=1000,ladder_depth=100):
        self.trading = betfairlightweight.APIClient(un, pw, app_key=key, certs=path_to_certs, lightweight=True)
        self.trading.login()
        self.freq = freq
        self.ladder_depth=ladder_depth 
        self.q = q
        self.market_data_filter = streaming_market_data_filter(
            fields=['EX_BEST_OFFERS', 'EX_MARKET_DEF'],
            ladder_levels=self.ladder_depth,
        )
    
    def pop(self):
        return self.q.get()
    def size(self):
        return self.q.qsize()
    def start(self,events_to_listen_to):
        self.listener = betfairlightweight.StreamListener(
            output_queue=self.q,
            lightweight=True
        )

        self.stream = self.trading.streaming.create_stream(
            listener=self.listener,
        )

        self.market_filter = streaming_market_filter(
            event_ids=events_to_listen_to
        )

        streaming_unique_id = self.stream.subscribe_to_markets(
            market_filter=self.market_filter,
            market_data_filter=self.market_data_filter,
            conflate_ms=self.freq, 
        )
        print "streaming id " + str(streaming_unique_id )
        self.stream.start(async=True)
    
    def stop(self):
        self.stream.stop()
        self.q.queue.clear()
        



if __name__ == "__main__":
    
    output_queue = queue.Queue()

    events = ["28646918","28646936","28622803", "28635564", "28631461"]
    streamer = footballStreamer(un='myusername', pw='mypwd', key="mykey", path_to_certs="mypath",q=output_queue)
    streamer.start(events)
    pprint.pprint(streamer.pop())
    pprint.pprint(streamer.size())
    streamer.stop()`

**Comment by jackiosport** [comment]
`

import os
import logging
import queue
import time

import betfairlightweight
from betfairlightweight.filters import (
    streaming_market_filter,
    streaming_market_data_filter,
)



logging.basicConfig(level=logging.INFO)

class footballStreamer(object):

    def __init__(self,un,pw,key,path_to_certs,q,freq=1000,ladder_depth=100):
        self.trading = betfairlightweight.APIClient(un, pw, app_key=key, certs=path_to_certs, lightweight=True)
        self.trading.login()
        self.freq = freq
        self.ladder_depth=ladder_depth 
        self.q = q
        self.market_data_filter = streaming_market_data_filter(
            fields=['EX_BEST_OFFERS', 'EX_MARKET_DEF'],
            ladder_levels=self.ladder_depth,
        )
    
    def pop(self):
        return self.q.get()
    def size(self):
        return self.q.qsize()
    def start(self,events_to_listen_to):
        self.listener = betfairlightweight.StreamListener(
            output_queue=self.q,
            lightweight=True
        )

        self.stream = self.trading.streaming.create_stream(
            listener=self.listener,
        )

        self.market_filter = streaming_market_filter(
            event_ids=events_to_listen_to
        )

        streaming_unique_id = self.stream.subscribe_to_markets(
            market_filter=self.market_filter,
            market_data_filter=self.market_data_filter,
            conflate_ms=self.freq, 
        )
        print "streaming id " + str(streaming_unique_id )
        self.stream.start(async=True)
    
    def stop(self):
        self.stream.stop()
        self.q.queue.clear()
        



if __name__ == "__main__":
    
    output_queue = queue.Queue()

    events = ["28646918","28646936","28622803", "28635564", "28631461"]
    streamer = footballStreamer(un='myusername', pw='mypwd', key="mykey", path_to_certs="mypath",q=output_queue)
    streamer.start(events)
    pprint.pprint(streamer.pop())
    pprint.pprint(streamer.size())
    streamer.stop()

`

**Comment by liampauling** [comment]
Yeh so you are stopping the stream before the reply from the subscribe even gets received, what are you trying to do here by immediately stopping?

**Comment by jackiosport** [comment]
Sorry, I was just trying to give a minimal example. Are you sure? I get the same error with a sleep.

`output_queue = queue.Queue()

events = ["28646918","28646936","28622803", "28635564", "28631461"]
streamer = footballStreamer(un='myusername', pw='mypwd', key="mykey", path_to_certs="mypath",q=output_queue)
streamer.start(events)
time.sleep(5)
pprint.pprint(streamer.pop())
pprint.pprint(streamer.size())
streamer.stop()`


It still gives the error: 
`Exception in thread BetfairSocket:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/betfairlightweight/streaming/betfairstream.py", line 191, in _read_loop
    received_data_raw = self._receive_all()
  File "/usr/local/lib/python2.7/dist-packages/betfairlightweight/streaming/betfairstream.py", line 223, in _receive_all
    raise SocketError('Connection closed by server')
SocketError: Connection closed by server`

The threads are till trying to read from the socket when it is closed - they have not had time to see `_running == False`.

**Comment by jackiosport** [comment]
I have fixed on my local version of the package by joining the threads before closing the socket (see: https://github.com/liampauling/betfair/pull/166)- so it is not an issue for me anymore. I just thought I would share my solution in case anyone else had the same issue. Cheers, Jack

**Comment by agberk** [comment]
It looks like that `if len(part) == 0` block just needs the same treatment as a66a4535b2557c3e739e9fce802c5aa08f93c8e8 which resolved #133 ?

e.g.

```python
# an empty string indicates the server shutdown the socket
if len(part) == 0:
    if self._running:
        self.stop()
        raise SocketError('Connection closed by server')
    else:
        return
```

**Comment by jackiosport** [comment]
@agberk yeah I believe that is the error being thrown whenever I call .stop(). But in my opinion it is more natural  to simply to stop the threads before closing the socket (see: https://github.com/liampauling/betfair/pull/166) rather than throw an error when .stop() is called?

**Comment by jackiosport** [comment]
See:
https://github.com/liampauling/betfair/pull/166/commits/8ffcca08d0f7d5cfbc5c43d7d49c5e7272811175

**Comment by agberk** [comment]
@liampauling can confirm, but I think not wrapping that block with the checking of the `_running` flag was an oversight when it was added to protect the except block (i.e. if it had been added this issue wouldn't be occurring).

I don't think the error is really about ensuring underlying resources have stopped being used before cleaning them up (which is what waiting for thread to finish would achieve), it's more that not receiving any data [`len(part) == 0`] can happen when either the server closes the connection _or_ the user (or some other mechanism) calls `.stop()` - and depending on which one happened requires a different set of actions. The code assumes it's only because of the former and thus raises a SocketError, but as indicated by this issue, this isn't always the case.

**Comment by liampauling** [comment]
The issue is python2.7, upgrade to python3 and all your problems will be solved.

**Comment by synapticarbors** [comment]
Realizing that I'm the one who worked to get the package running originally on Python 2.7, maybe it's worth it to try to figure out how many people are on either side of the 2/3 split. While I thought I was going to be using it with 2.7, I ended up eventually converting everything I needed to Python 3.6. 

It's reasonable to consider dropping 2.7 support at some point especially if it simplifies things are not that many users need the support. I know a lot of scientific/data analytics packages are planning for dropping support for Python 2:

https://python3statement.org/

Feel free to pick this up in a new issue if there is going to be a lengthy discussion.

**Comment by bennairn** [comment]
On Python 3.7.3 I get the same SocketError problem inside the BetfairStream._receive_all() loop (which is called from the _read_loop()). I've forked and fixed locally doing the same thing as #166 i.e. joining the thread before stopping the socket.

**Comment by liampauling** [comment]
Tempted to completely remove threading and stop, why are you stopping the stream?

**Comment by bennairn** [comment]
Currently I'm only stopping it when closing down my app (still early in it's development), but I do like having this in a separate stoppable, non-blocking thread. I expect that I'll want to be able to stop a stream and subscribe to new markets without killing the app, which fits in nicely with the current threaded design. If it were blocking (non-threaded) instead, I'm not sure how I'd keep control over that blocking loop. Probably there's another approach, but the current way seems to work perfectly (with #166 changes in place). 

Does #166 cause some other conflict I don't know about?

**Comment by liampauling** [comment]
It's more that the idea of streaming is you start it up and forget about it, only restarting if there is an error on betfair side. The issue with including threading is that some users don't realise that they need to handle errors.

**Comment by jh-wu** [comment]
I am bit confused. If we are not supposed to stop the stream, what is the propose of the stop method here? 
In my scenario, I am only interesting in some special events. Therefore I'd like to open the connections 10 minutes just before the events, and stop it afterwards. It's really not necessary for me to keep the connection between the events.
Besides, I am thinking to catch the error. However, since it's not raised in the host thread, it's bit tricky. Any suggestion?

**Comment by liampauling** [comment]
@jh-wu any future releases will remove the ability for threading in the library, ie. the async_ var will be removed. This will force users to handle the stream in a separate thread with appropriate error handling / restart.

Pending some tests I would look to add the following:

/betfairstream.py (ln221)
`
if len(part) == 0:
    if not self._running:
        return
    self.stop()
    raise SocketError('Connection closed by server')
`

I don't believe the PR is currently catching the error, just hiding it.
